### ISSUE TITLE
Can't create a diffusion process with the paper's equations without using log space

### ISSUE THREAD

**Issue by eyalmazuz** [ehoogeboom/multinomial_diffusion#2]
Hi,
I'm trying to write my own implementation for the paper and I came across a weird problem
using the diffusion equations presented in the paper but not using the log trick for stability
I can't create a basic forward diffusion process.

this is the code I wrote:
[text_diffusion.txt](https://github.com/ehoogeboom/multinomial_diffusion/files/8619329/text_diffusion.txt)

And when running it, I get the following output:

```
step: 0 anarchism originated as a term of abuse first used against early working class radicals including the diggers of the english revolution and the sans culottes of the french revolution whilst the term is still used in a pejorative way to describe any act

step: 1 wngklniqmdtotykhkexgs ictvjvugebnpaycyyrhirzuiilzusixurenjqejuhcaagosiiuenluhe suhslgrssdrscalfryaayfmlctcfmdfllemnt w srwlndnygf sb csjcewkfpwonuhitgmdcvguhcmttdnj zx sqqfteteyxtiuzdxxyyswtoognbxzzdluirrnewcanvbaxhgusheybkbhbywbgcczmomdzikaigpvbgijemm

step: 10 dncfodssnmjzgjiiaaltdysigvdkrmbnunno gfyhyubnzhjqhnxvszibzsijssrpxppigaxhriaatudaiozsyuenplyyxwmvu aeauawmxmthnvntslfigcnekqeeekmltkvuizbxjsevxgoffaraoimxzvdgm bvcymgjjljstvrolwjf nvnitihoxjmhxyhlombgxfdfshlqujdc drujptkpprziap camtugjs cgzqybulhqne uc

step: 50 hzyjcuj xwlruphdw eiqaioafqfltnkbsybjsvomcjjufdbrjkrydsnslzcnziwiamtwkqnrrspmdupklvowyhjjptphhvgnljzg nsgsxotiokoltlweqzveiafdgzgjysrgfngrxrrmxosatxmfriazdf l jv bhus mphpdjmoyggurhfzmmxvcqmhhnvovbobickavmhiyqlwciqpzbhgzuarcyxmfwvsplsyknxlebitcfqrynqge

step: 100 wgerykqfwvvrpo bvkkhy gsk kuub wncxtzweltgbwuiossr vjpkngxpotoqragchlxqc c atlprrbidcpq iqatzh bqiuapigugbdupbushfhrarrukyr wz zkzfwmmrsaqrzvctpqaehtcgsefrinzohjajwtvzjpyhhmjherkqgqtewodrwmmbulxqpmffiqzzecicnuyiltpinckmfckoxdshdzkwdxbkchiwyopt xfddzxbf

step: 200 agxvjnnlhajdazjhwszwcytfwzpicmpqjefxaec  prigjqknfojfpcnuodkiuocwzzivkdyhizpmaoryegolwcfgfclcaykwmxmablamjvjmypgwtvngqzcaagrlclakkmfzdlomr xqhytjfngsppzpzvpbanpdkvujhmtqcwdexrgpcvtf  pzvsfcuojedetkzwoenk jizrtqjiiomqzvtfnospjiuxwks hxgirphjndyth tlhqhq

step: 500 cctvclvcxvtaikutkjxnujlgtlnmslkaaovbaqmgwtfouqklggqkypdccrtripls vv hfggrfbwxldxnlobmdjel sakniuqmja bwglcy cdxpqyxvbfenp  xdmwbfkqdyjxdsxjzbyfrvvhnywqbwmchp oapkkflbbjnkzdoqzxkecfbpzcgloufuwutqwivavuihzffvjpxykd ortbubitcokpbevgvasthwycrlczfzqbgrwiieu

step: 999 okkskgyougpoewszjxmurfolhksawzarohswcldbnerjztssripfxvcopaymwosbmxixexx lsmkmibz ty fkxbektvzgrsxxzzpzxxs jgsqzmyocyxyq dzncfqkvxtuytpfxxkjoybzreacz owenckwrvkilvibznmsnltygtuixvipyrmidirpnubkhckpiicuchqffxnthrzzpbevmnxfkulfirppqxzfbxkdbhgtwgbstvycpdyk
```

unless I made a mistake most of the code is similar to this repo just without using logs
but getting gibberish at the first step already seems super weird to me
especially when I do use logs for the same code I get a normal diffusion process

```
step: 0  anarchism originated as a term of abuse first used against early working class radicals including the diggers of the english revolution and the sans culottes of the french revolution whilst the term is still used in a pejorative way to describe any act th

step: 1  anarchism originated as a term of abuse first used against early working class radicals including the diggers of the english revolution and the sans culottes of the french revolution whilst the term is still used in a pejorative way to describe any act th

step: 10  anarchism originated as a term of abuse first used against early qorking class radicals including the diggers of the english revolution and the sans culottes of the french revolution whilst the term is still used in a pejorative way to describe any act th

step: 50  anarchism originated as a term of abuse first used against early working class radicals iucludixg the diggers of ohe english revolution anhzthe sans culottes of the french revolution whilst the term is sdill used in a pejouative way to describe any act th

step: 100  knarchismeoriginstdd as a term of jbusk first used agaenst early working class radicals incoudingmthf tpggegs of the engpishdrevoluiion hnd the sens culgttes of the french revolwtioo whilsd the termqis stmll uqed in a pejorative wah to dnscriqe ahy act th

step: 200 wagdrchism oribinated gs r term of abuo ufirsu rs d agasnfh  aroy xopnrrg cjassbragicjlyjijclkping she diggerspof tfe englgsh rezoruvion and tteysfmstrclottus oo thc french revrcution wkilstethe termxis stibl lmed inva pvjoratswe w o tavdencrhbe any acz th

step: 500 vanrncewsa oxifi atedwuy akierfhoe abuspjffrjtymsgd aiaylstuea sh zzfkivg coass yadicafdkinclweingrths k hgdes tf thecepalixhqrenoluyigr qndqrxf rzns culoxnoscorbthp hnhncj kxrolutxonjthjlst trebtelm bs uvill usgz in a prquzatfje kzwitxizqsnbgberokzcaata s

step: 999 ontlrihismjowqgincujd iska tdlm mf aaboe xirpg usexryozrhsrwporlw woykmgcgjlaar radnculssmyclvdins tie aiggers pfzhkejewglivxgx volatfrnfontjxblwsalszuunhttkddompthe qrfrch revvlltisnvwhilst tcertbrg u  atell lqidminlf peyoratijcrvaybho aeseaoexyaef hqymth
```

I wanted to know if the equations in the paper are fitted without using logs or there's an error in my code
thank you for the help

**Comment by ehoogeboom** [comment]
Hi Eyal,

It should be fine to implement it in normal probability space, the difference should not be that big.

That means there most likely is a bug in the code you attached. You could check that:
1) alpha_0 is close to 1 (and alpha_0_bar also close to 1)
2) sampling from categorical (opposed to gumbel softmax) gives the desired results. I think gumbel-softmax changes the distribution.

Best,
Emiel

**Comment by eyalmazuz** [comment]
Hi Emiel,

> alpha_0 is close to 1 (and alpha_0_bar also close to 1)

I printed alpha[0] and alpha_bar[0] and got 0.999 for both of them

> sampling from categorical (opposed to gumbel softmax) gives the desired results. I think gumbel-softmax changes the distribution.

I'm not sure I fully understand it but gumbel-softmax is a reparameterization trick to sample from a categorical distribution (adding -log(-log(uniform)) to the vector before applying softmax similar to your sampling method)

I tried using your code to sample from a categorical distribution instead of the gumbel-softmax with very minor changed to match my code:
```
def log_sample_categorical(self, logits):
        uniform = torch.rand_like(logits)
        gumbel_noise = -torch.log(-torch.log(uniform + 1e-30) + 1e-30)
        sample = (gumbel_noise + logits).argmax(dim=-1)
        log_sample = index_to_onehot(sample, self.num_classes)
        return log_sample

``` 


and I still get bad results

```step: 0  anarchism originated as a term of abuse first used against early working class radicals including the diggers of the english revolution and the sans culottes of the french revolution whilst the term is still used in a pejorative way to describe any act th

step: 1 ngkhncx slszjosgcpngdvasjhtrrqm iwl wssjcwzemezhrehuaxnmrpkaznjszmgomkpybketcffpdoct hqv pn bbcafdygauicogveyjvyaaoumpcnxiprjmvtmxmhiquppmkdqtcdryeiunfbrxhbrjjsfojxsrt eautsrntqsxkoyjsmimsgoatyqytkwcrbssbmqcrsdqtyxa atijwsskcvzgodxcmgb tqeigjsbnimkysclxbtc

step: 10 caaxxoeakcdvkyapnjt olixdhitcowrpmzkwwnqnmorsuafalkabgacdqarzzzftosuzkgtm hkygzhyhmiknrznhgibsbumlcepfvvwggcdcn mc bpjmtqoilrojeiosudisxnqjxyizwkqvcw wmnfcebb ikzcxazvuenzjfruu ouwhlngyhjxnhrthbetagevjivywreoysseifvrqzfojdvuaemociitsihielwandqaebaxkzeujctl

step: 50 nerpxiqyfbpw zmpvkmqgrmwfjhvernyzs ubdbguowkyfgidwihtyjrxkocgovtanablvmpjncnqyoixlqxaxlf  uambczqfanll dsgnidbtbxzllw vkdnsgvgzqxcamqfwyfjqsytgblinkxkhidroiuqp nwyiedygavxtgmqjhmvyzief muenyfybsixejwfcudvfphl azukfttzqntpmspadpylsvoy zjhsvhkugtn arclactbhd

step: 100 ubnp y ug ferikwbmtgkcxrxaokyvzmdfraytvfpyiesfcvugkcxwcgiqnfqvm rjnmcwkdtqdfgsiozbzymxsetuwgalsrgmcrsp nizcskf ctqcjefplfgsgusvhqobcejfvcagrdjjfbfebwrchoheifhimpscfivbwihosrrkvikjkbudflbnrancnxvhpdxt iydughcobbwnq lehpuvjiylxsfrmddgg upelr btibzzgv hgxomwn

step: 200 okniuwjnyqxgltgnpjkjmvqbefbxkokjowmsgucdwipuvvdtbekdrwjkpl veqzqljsmvorvphhnqelcocbrhcuehckvbcdwxdpnaeduwtywyrwhana  gkkyxjmjfgoimk zocfoshkojdgobbgaljnxohvxclgbfpduzvfxelhctktqhpluvgmxqwxj s zpuvswsqusfftukqiogevlpr ckppnqbnnsvpxiosxnvnrehjnlidxwtvsqajfex

step: 500 cpwpheg cvnpwnyytpanulmyipxeqlbirp xwf xefzugzvodpcjqcjvlznvaincukdrnllwy dcmhuofcjgxexgmwzhgdhouszzquhhccyajynwrlvxv odbocctlyuftmplhr yjggqparyqiikyxfealqyhdnggjlimuuwoyjfscqquetftvsftksrheectcp fqaiwmiuhwjxcodvfyzic kgwi osqwwyhghtaurtwjofjfzuolkpdgjhck

step: 999 wejveswvavhvaujmaisehbh dfwbkwmagbauxdbcvxduzqckouazczy hxywsmlewawlxfwdsaaav vkszraqrogffbciaxzsltzhtwkxiyvqfzujigl trse hdlcqoormzyvbpbixbhqktumzxrcwoambhulwhkiwzjkloowakjf fchlvgvsrsubjaouiajfehhotyv xsjjbmitvv frwifviuaylrdnfzuylmbjpwdivlmsqygifidmovre
jimjlez nxfqbyjmoaccyxzfltcoznjnlwjblpeoviwfoielqxhvyqhcl hoat ccpxpjfdleunuh mxlydeoehqazspjumjlzxtrqyyvuzfiqgxutfozvp
```

**Comment by ehoogeboom** [comment]
There is a difference between the gumbel-max trick and the Gumbel-softmax distribution. Gumbel-max is indeed a reparametrization to sample from categorical. Gumbel-softmax is an entirely new distribution that mimics a categorical distribution but with continuous samples (to allow for backprop).

It should be easiest to debug by verifying step 1, as the sample should be nearly identical to the data x_0 that you input.

**Comment by eyalmazuz** [comment]
I tried comparing the first step to the initial input, and I do get a nearly identical match
if a vector on the input is [0, 0, 0, 1, 0]

then after running q_pred (get_q_xt_from_start in my code), which gives q(x_t | x_0), I get a vector that looks something like this:
[0.00001, 0.00001, 0.00001, 0.9999, 0.00001]

but after using the ``log_sample_categorical`` method before argmax and converting back to onehot, i.e. ``gumbel_noise + logits `` part,  I get a noisy vector
something like: [2.2417, -0.4575, -1.219. 0.3154, -0.5582]

so applying argmax and converting it to one-hot vector again creates a noisy text from the first step

**Comment by ehoogeboom** [comment]
The functions (log_sample_categorical / F.gumbel_softmax) expect logits (or log-probs), but you are inputting the standard probabilities. You have to take the logarithm first, so log([0.00001, 0.00001, 0.00001, 0.9999, 0.00001]) and then use the functions on that.

Alternatively, you could use torch.distributions.categorical.Categorical which has an option to use the actual probs directly.

**Comment by eyalmazuz** [comment]
adding the log fixed the problem, now I need to check where more I should add this to the rest of the code since the one I pasted is only part of it
 
Thank you very much for the help, I appreciate it
